nav2_costmap_2d/src/costmap_subscriber.cpp

```diff
@@ -16,41 +16,82 @@
 #include <memory>
 
 #include "nav2_costmap_2d/costmap_subscriber.hpp"
+#include "nav2_costmap_2d/cost_values.hpp"
+
+using nav2_costmap_2d::NO_INFORMATION;
+using nav2_costmap_2d::LETHAL_OBSTACLE;
+using nav2_costmap_2d::FREE_SPACE;
 
 namespace nav2_costmap_2d
 {
 
 CostmapSubscriber::CostmapSubscriber(
   nav2_util::LifecycleNode::SharedPtr node,
-  const std::string & topic_name)
+  const std::string & topic_name,
+  bool use_raw)
 : CostmapSubscriber(node->get_node_base_interface(),
     node->get_node_topics_interface(),
     node->get_node_logging_interface(),
-    topic_name)
+    node->get_node_parameters_interface(),
+    topic_name,
+    use_raw)
 {}
 
 CostmapSubscriber::CostmapSubscriber(
   rclcpp::Node::SharedPtr node,
-  const std::string & topic_name)
+  const std::string & topic_name,
+  bool use_raw)
 : CostmapSubscriber(node->get_node_base_interface(),
     node->get_node_topics_interface(),
     node->get_node_logging_interface(),
-    topic_name)
+    node->get_node_parameters_interface(),
+    topic_name,
+    use_raw)
 {}
 
 CostmapSubscriber::CostmapSubscriber(
   const rclcpp::node_interfaces::NodeBaseInterface::SharedPtr node_base,
   const rclcpp::node_interfaces::NodeTopicsInterface::SharedPtr node_topics,
   const rclcpp::node_interfaces::NodeLoggingInterface::SharedPtr node_logging,
-  const std::string & topic_name)
+  const rclcpp::node_interfaces::NodeParametersInterface::SharedPtr node_parameters,
+  const std::string & topic_name,
+  bool use_raw)
 : node_base_(node_base),
   node_topics_(node_topics),
   node_logging_(node_logging),
-  topic_name_(topic_name)
+  node_parameters_(node_parameters),
+  topic_name_(topic_name),
+  use_raw_(use_raw)
 {
-  costmap_sub_ = rclcpp::create_subscription<nav2_msgs::msg::Costmap>(node_topics_, topic_name_,
+  if (use_raw_) {
+    costmap_raw_sub_ = rclcpp::create_subscription<nav2_msgs::msg::Costmap>(node_topics_,
+        topic_name_,
+        rclcpp::QoS(rclcpp::KeepLast(1)).transient_local().reliable(),
```

Comment:
You'd have to tell me - is this the right QoS for a costmap? I would think you'd rather have the most recent on the raw costmaps than something old. For the normal costmap I think that's necessary if using the updates topic since the updates need a correct base map.

Reply:
I agree that this doesn't seem correct. I don't think any of these topics (including OccGrid and OccGridUpdate) should be transient_local - that's for messages that get published infrequently. 

I'd get rid of the reliable on the raw topic  as well.

Reply:
Currently, the `Costmap2DPublisher` uses these settings when publishing all the costmap topics.
https://github.com/ros-planning/navigation2/blob/abe51671084bcc18dd210e95edab962b92d518eb/nav2_costmap_2d/src/costmap_2d_publisher.cpp#L59